sw/device/lib/dif/dif_adc_ctrl.c

```diff
@@ -0,0 +1,173 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_adc_ctrl.h"
+
+#include <assert.h>
+
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/lib/dif/dif_base.h"
+
+#include "adc_ctrl_regs.h"  // Generated.
+
+static_assert(ADC_CTRL_PARAM_NUM_ADC_CHANNEL == 2,
+              "Expected two ADC Controller channels.");
+static_assert(ADC_CTRL_PARAM_NUM_ADC_FILTER == 8,
+              "Expected eight ADC Controller filters.");
+static_assert(ADC_CTRL_ADC_CHN0_FILTER_CTL_MIN_V_FIELD_WIDTH == 10,
+              "Expected channel-0 filter min-voltage field to be 10 bits.");
+static_assert(ADC_CTRL_ADC_CHN1_FILTER_CTL_MIN_V_FIELD_WIDTH == 10,
+              "Expected channel-1 filter min-voltage field to be 10 bits.");
+static_assert(ADC_CTRL_ADC_CHN0_FILTER_CTL_MAX_V_FIELD_WIDTH == 10,
+              "Expected channel-0 filter max-voltage field to be 10 bits.");
+static_assert(ADC_CTRL_ADC_CHN1_FILTER_CTL_MAX_V_FIELD_WIDTH == 10,
+              "Expected channel-1 filter max-voltage field to be 10 bits.");
+
+dif_result_t dif_adc_ctrl_configure(const dif_adc_ctrl_t *adc_ctrl,
+                                    dif_adc_ctrl_config_t config) {
+  if (adc_ctrl == NULL ||
+      config.power_up_time_aon_cycles > ADC_CTRL_ADC_PD_CTL_PWRUP_TIME_MASK ||
+      config.wake_up_time_aon_cycles > ADC_CTRL_ADC_PD_CTL_WAKEUP_TIME_MASK) {
+    return kDifBadArg;
+  }
+
+  uint32_t en_ctrl_reg = 0;
+  uint32_t pd_ctrl_reg = 0;
+  uint32_t lp_sample_ctrl_reg = 0;
+  uint32_t np_sample_ctrl_reg = 0;
+
+  switch (config.mode) {
+    case kDifAdcCtrlLowPowerScanMode:
+      pd_ctrl_reg = bitfield_bit32_write(pd_ctrl_reg,
+                                         ADC_CTRL_ADC_PD_CTL_LP_MODE_BIT, true);
+      pd_ctrl_reg = bitfield_field32_write(pd_ctrl_reg,
+                                           ADC_CTRL_ADC_PD_CTL_PWRUP_TIME_FIELD,
+                                           config.power_up_time_aon_cycles);
```

Comment:
Does `power_up_time_aon_cycles` also need a little blurb in the struct's doc about only applying to Low Power Scan Mode?